clients/java/src/test/java/io/zeebe/client/impl/worker/JobWorkerImplTest.java

```diff
@@ -0,0 +1,238 @@
+  private static final Duration SLOW_POLL_THRESHOLD = Duration.ofMillis(SLOW_POLL_DELAY_IN_MS / 2);
+
+  @Rule public final GrpcCleanupRule grpcCleanup = new GrpcCleanupRule();
+
+  private MockedGateway gateway;
+  private ZeebeClient client;
+
+  @Before
+  public void setup() throws IOException {
+    // Generate a unique in-process server name.
+    final String serverName = InProcessServerBuilder.generateName();
+
+    // Create a server, add service, start, and register for automatic graceful shutdown.
+    gateway = new MockedGateway();
+    grpcCleanup.register(
+        InProcessServerBuilder.forName(serverName)
+            .directExecutor()
+            .addService(gateway)
+            .build()
+            .start());
+
+    // Create a client channel and register for automatic graceful shutdown.
+    final ManagedChannel channel =
+        grpcCleanup.register(InProcessChannelBuilder.forName(serverName).directExecutor().build());
+
+    // Create the Zeebe Client
+    client =
+        new ZeebeClientImpl(new ZeebeClientBuilderImpl(), channel, GatewayGrpc.newStub(channel));
+  }
+
+  @Test
+  public void shouldBackoffWhenGatewayRespondsWithResourceExhausted() {
+    // given a gateway that responds with some jobs
+    gateway.respondWith(TestData.jobs(10));
+
+    // and a client with retry delay supplier that is slowing down polling
+    client
+        .newWorker()
+        .jobType("test")
+        .handler(NOOP_JOB_HANDLER)
+        .backoffSupplier(prev -> SLOW_POLL_DELAY_IN_MS)
+        .open();
+
+    // and assuming that the gateway responded multiple times successfully with jobs
+    gateway.startMeasuring();
+    Awaitility.await()
+        .pollInterval(Duration.ofMillis(10))
+        .atMost(Duration.ofSeconds(1))
+        .until(() -> gateway.getCountedPolls() > 3);
+    gateway.stopMeasuring();
+
+    // then polling is fast
+    assertThat(gateway.getTimeBetweenLatestPolls()).isLessThan(SLOW_POLL_THRESHOLD);
+
+    // when the gateway responds with errors
+    gateway.respondWith(new StatusRuntimeException(Status.RESOURCE_EXHAUSTED));
+
+    // then polling is slowed down
+    gateway.startMeasuring();
+    Awaitility.await()
+        .pollInterval(Duration.ofMillis(100))
+        .atMost(Duration.ofSeconds(5))
+        .untilAsserted(
+            () ->
+                assertThat(gateway.getTimeBetweenLatestPolls()).isGreaterThan(SLOW_POLL_THRESHOLD));
+  }
+
+  /**
+   * This mocked gateway is able to record metrics on polling for new jobs and easily switch how it
+   * responds to polling.
+   *
+   * <ul>
+   *   Due to the concurrent nature of the test setup and the job worker, 2 lock objects are used:
+   *   <li>responsesLock to lock access to the mocking of responses objects for test setup;
+   *   <li>metricsLock to lock access to the polling metrics objects.
+   * </ul>
+   */
+  private static final class MockedGateway extends GatewayImplBase {
+
+    // mocking of poll responses
```

Comment:
I understand the comment groups state by purpose, can you explain what it adds in terms of comprehension that I wouldn't get from the field names?

Reply:
Just deepens grouping for me, which is related to the multiple responsibilities of this class. It made this separation more clear to me, but I can also remove it if you think that's better.

Reply:
I would remove it, as I don't think it serves anything, other than potentially be confusing when someone inevitably adds a field which has nothing to do with it :smile: